tiledb/sm/serialization/query.cc

```diff
@@ -923,6 +908,54 @@ Status query_deserialize(
   STATS_FUNC_OUT(serialization_query_deserialize);
 }
 
+Status query_deserialize(
+    const Buffer& serialized_buffer,
+    SerializationType serialize_type,
+    bool clientside,
+    CopyState* copy_state,
+    Query* query) {
+  // Create an original, serialized copy of the 'query' that we will revert
+  // to if we are unable to deserialize 'serialized_buffer'.
+  BufferList original_bufferlist;
+  Status st =
+      query_serialize(query, serialize_type, clientside, &original_bufferlist);
+  RETURN_NOT_OK(st)
+
+  // The first buffer is always the serialized Query object.
+  tiledb::sm::Buffer* original_buffer;
+  st = original_bufferlist.get_buffer(0, &original_buffer);
+  RETURN_NOT_OK(st)
+
+  // Similarly, we must create a copy of 'copy_state'.
+  std::shared_ptr<CopyState> original_copy_state = nullptr;
```

Comment:
Gotcha. Could you use a `unique_ptr` here? Probably a better default when possible.

Reply:
Yeah, good idea. Will do.

Reply:
Changed, but not without a pedantic caveat:

I chose to use a `shared_ptr` over a raw pointer for obvious reasons. However, I chose a `shared_ptr` over a `unique_ptr` because while the associated constructor `make_shared` is defined in C++11, `make_unique` is not in the standard until C++14. I placed an implementation of `make_unique` in the `tiledb::sm::utils` namespace. I saw that all other definitions in `tiledb/sm/misc/utils.h` are scoped inside a nested namespace within `tiledb::sm:utils`, e.g. `tiledb::sm::utils::datatype` and `tiledb::sm::utils::time`. I wasn't sure where to stick this one, so I left it in `utils`.

Reply:
Will this still build if someone is using a C++14-enabled toolchain?

Reply:
Yeah, definitely. Good question. There won't be any sort of signature collision since they're in different namespaces.

Reply:
Is there a reason to not just use the `unique_ptr` from raw pointer constructor in place instead of make_unique function?

Reply:
As a best practice to avoid an extra memory allocation and to avoid ever holding a raw pointer.

Reply:
Note that if it this was a one-off requirement, I would have just used `unique_ptr` to construct it. However, we probably will get re-use out of this for future `unique_ptr` objects.

Reply:
@tdenniston @Shelnutt2 I scraped the custom `make_unique` because I was having to wrestle with the visual studio compiler over template syntax. Another day, perhaps.